src/Database/LSMTree/Internal/RawOverflowPage.hs

```diff
@@ -0,0 +1,162 @@
+-- RawOverflowPage type
+-------------------------------------------------------------------------------
+
+-- | When a key\/op pair is too large to fit in a single disk page, the
+-- representation is split into a normal page, and one or more overflow pages.
+-- The normal 'RawPage' follows the run page format, and contains the key,
+-- optional blob reference and a prefix of the value, while the overflow pages
+-- contain the suffix of a large value that didn't fit in the normal page.
+--
+-- Each overflow page is the same size as normal pages (currently 4096 only).
+--
+data RawOverflowPage = RawOverflowPage
+                         !Int        -- ^ offset in Word8s.
+                         !ByteArray
+  deriving (Show)
+
+-- | This invariant is the same as for 'RawPage', but there is no alignment
+-- constraint. This is for two reasons: 1. we don't need alignment, because
+-- the page has no structure that we need to read with aligned memory
+-- operations; 2. we don't want alignment because we want to convert the suffix
+-- of serialised values (as 'RawBytes') into a 'RawOverflowPage' without
+-- copying, and a suffix can start at an arbitrary offset.
+--
+invariant :: RawOverflowPage -> Bool
+invariant (RawOverflowPage off ba) = and
+    [ off >= 0                            -- offset is positive
+                                          -- no alignment constraint
+    , (sizeofByteArray ba - off) >= 4096  -- there are always 4096 bytes
+    , isByteArrayPinned ba                -- bytearray must be pinned
+    ]
+
+instance NFData RawOverflowPage where
+  rnf (RawOverflowPage _ _) = ()
+
+-- | This instance assumes pages are 4096 bytes in size
+instance Eq RawOverflowPage where
+    RawOverflowPage off1 ba1 == RawOverflowPage off2 ba2 =
+      P.Vector off1 4096 ba1 == (P.Vector off2 4096 ba2 :: P.Vector Word8)
+
+-- | Create a 'RawOverflowPage'.
+--
+-- This function may copy data to satisfy internal 'RawOverflowPage' invariants.
+-- Use 'unsafeMakeRawOverflowPage' if you don't want copy.
+--
+makeRawOverflowPage
+    :: ByteArray  -- ^ bytearray
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes
+    -> RawOverflowPage
+makeRawOverflowPage ba off len
+    | len >= 4096
+    , let page = RawOverflowPage off ba
+    , invariant page
+    = page
+
+    | otherwise
+    = makeRawOverflowPageCopy ba off len
+
+makeRawOverflowPageCopy
+    :: ByteArray  -- ^ bytearray
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes
+    -> RawOverflowPage
+makeRawOverflowPageCopy ba off len =
+    (\page -> assert (invariant page) page) $
+    RawOverflowPage 0 $ runByteArray $ do
+      mba <- newPinnedByteArray 4096
+      let suffixlen = clamp 0 4096 (min len (sizeofByteArray ba - off))
+      copyByteArray mba 0 ba off suffixlen
+      when (suffixlen < 4096) $ fillByteArray mba suffixlen 4096 0
```

Comment:
The second argument to `fillByteArray` is *number of bytes to fill*, not the end offset.

Reply:
I think we should enable  https://downloads.haskell.org/ghc/latest/docs/users_guide/debugging.html?highlight=bounds#ghc-flag--fcheck-prim-bounds on our packages, `vector` and `primitive`.

Reply:
> The case where len is larger than the remaining byte array is a bit suspicious to me. 

It's not. If `ByteArray` is unpinned, we must copy, as our `RawPage*`s have an invariant of being always pinned.

Reply:
> The second argument to fillByteArray is number of bytes to fill, not the end offset.

I am aware. And it's clear that we want to copy at most `len` bytes and we also *can't* copy more bytes than `sizeofByteArray ba - off`. The usage of `min` still seems suspicious to me.
If someone passes in a `len` that is larger than the byte array allows, that seems like a mistake from the caller to me (unless there is a legitimate use case, which I asked for). And if it's a mistake, I'd rather fail instead of silently truncating the copy and pretending to have built a page with `len` user-specified bytes.

Reply:
> If someone passes in a len that is larger than the byte array allows

This is an internal function.

We could add an `assert`.

Reply:
... but then we should also assert that `off` is `>=0` and `<= sizeofByteArray ba` and ... i.e. `assert (isValidSlice ba off len)`.

Reply:
Sounds reasonable to me.

Reply:
I also just noticed that I'd got the fill args wrong.

I agree that enabling `-fcheck-prim-bounds` in CI is a good idea.

Reply:
Note that the `primitive` package's `setByteArray` and `fillByteArray` do not currently use the ghc primop `setByteArray#` and so do not benefit from `-fcheck-prim-bounds`.

Reply:
I *think* this is now resolved.